nexus/sdk-python.md

```diff
@@ -0,0 +1,330 @@
+
+```python
+# This code (from Python gRPC docs) is not part of the proposal
+class MyService:  # instantiated at server-start time
+    def my_handler(self, my_input, context) -> MyResult: ...
+```
+
+However, unlike traditional RPC/HTTP handlers, a Nexus operation represents a long-running execution and must support three additional verbs (`get_info`, `get_result`, `cancel`), and must support `start` returning both synchronously and asynchronously (including making that decision at request-handling time).
+
+## 2a. Fully manual class-based operation definition
+
+The “fully manual” way we will support this in Nexus Python is that a user will define a class for the service, and a class for each operation.
+
+**[ALTERNATIVES CONSIDERED]** None. Giving the user the option to define classes at both levels for maximum implementation flexibility is the natural choice for Python (as it was in Java).
+
+```python
+import nexusrpc.handler
+
+@nexusrpc.handler.service(interface=interface.MyNexusService, name="my-service")  # import-time check that interface was implemented
+class MyNexusService: # instantiated when instantiating worker
+    # User is free to define custom constructor
+
+    @nexusrpc.handler.operation(name="my-invalid-python-identifier")
+    def echo(self) -> nexusrpc.handler.Operation[EchoInput, EchoOutput]:  # typing.Protocol interface
+        # Factory called when instantiating worker
+        return EchoOperation()  # User can pass anything to operation constructor
+
+class EchoOperation: # interface declaration optional; see below
+    # User is free to define custom constructor
+
+    async def start(self, input: EchoInput, options: nexusrpc.handler.StartOperationOptions) -> EchoOutput:
+        return EchoOutput(input.message)
+
+    async def fetch_info(self, token: str, options: nexusrpc.handler.FetchOperationInfoOptions) -> nexusrpc.handler.OperationInfo:
+        raise NotImplementedError
+
+    async def fetch_result(self, token: str, options: nexusrpc.handler.FetchOperationResultOptions) -> EchoOutput:
+        raise NotImplementedError
+
+    async def cancel(self, token: str, options: nexusrpc.handler.CancelOperationOptions) -> None:
+        raise NotImplementedError
+```
+
+- User chooses whether to implement any verb other than `start`. When `start` always returns synchronously (as here), the other verbs will typically not be used, but the spec doesn’t rule it out.
+
+Here’s what Temporal worker start-time might look like:
+
+```python
+worker = Worker(
+    client,
+    task_queue=task_queue,
+    workflows=[HelloWorkflow],
+    # User has defined custom service and operation constructors,
+    # passing the service instance or the db client through
+    # to give verb handlers access to this db client
+    nexus_services=[MyNexusService(db_client=MyDBClient.connect(...))],
```

Comment:
Same concerns here with reusing the name service to mean service impl. It can be confusing to a user whether they should pass in the service or the service (the sentence is ambiguous on purpose to convey confusion when discussing Nexus).